updated event hubs migration guide to follow template (#15953)
Azure/azure-sdk-for-python: sdk/eventhub/azure-eventhub/migration_guide.md

# Guide for migrating azure-eventhub to v5 from v1

This guide is intended to assist in the migration to `azure-eventhub` v5 from v1. It will focus on side-by-side comparisons for similar operations between the two packages.

Familiarity with the `azure-eventhub` v1 package is assumed. For those new to the Event Hubs client library for Python, please refer to the [README for `azure-eventhub`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/eventhub/azure-eventhub/README.md) rather than this guide.

## Table of contents

* [Migration benefits](#migration-benefits)
    - [Cross Service SDK improvements](#cross-service-sdk-improvements)
    - [New features](#new-features)
* [Important changes](#important-changes)
    - [Client hierarchy](#client-hierarchy)
    - [Client constructors](#client-constructors)

>     - [Sending](#sending-events)
>     - [Receiving](#receiving-events)

Reviewer: nit:
```suggestion
    - [Sending events](#sending-events)
    - [Receiving events](#receiving-events)
```

We've also merged the functionality from `EventProcessorHost` into `EventHubConsumerClient`. 

#### Approachability
By having a single entry point for sending, the `EventHubProducerClient` helps with the discoverability of the API
as you can explore all available features for sending events through methods from a single client, as opposed to searching
through documentation or exploring namespace for the types that you can instantiate.

Similarly, by having a single entry point for receiving of any type (from single partition, all partitions, or with load balancing and checkpointing features) within Event Hubs, the `EventHubConsumerClient` helps with the discoverability of the API as you can explore all available features for receiving events through methods from a single client, as opposed to searching
through documentation or exploring namespace for the types that you can instantiate.

#### Consistency
We now have methods with similar names, signature and location for sending and receiving.
This provides consistency and predictability on the various features of the library.

### Client constructors

> - While we continue to support connection strings when constructing a client, the main difference is when using Azure Active Directory.
> We now use the new [`azure-identity`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md) library
> to share a single authentication solution between clients of different Azure services.

Reviewer 2: ```suggestion
While we continue to support connection strings when constructing a client, below are the differences in the two versions:
- In v5, we now support the use of Azure Active Directory for authentication. 
The new [`azure-identity`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md) library allows us 
to share a single authentication solution between clients of different Azure services.
- The option to construct a client using an address of the form `amqps://<SAS-policy-name>:<SAS-key>@<fully-qualified-namespace>/<eventhub-name>` is no longer supported in v5. This address is not readily available in the Azure portal or in any tooling and so was subject to human error. We instead recommend using the connection string if you want to use a SAS policy.
```

- In v1, the `receive` method returned a list of `EventData`. You would call this method repeatedly every time you want receive a set of events. In v5, the new `receive` method takes user callback to process events and any resulting errors. This way, you call the method once and it continues to process incoming events until you stop it.
- Additionally, we have a method `receive_batch` which behaves the same as `receive`, but calls the user callback with a batch of events instead of single events. 
- The same methods can be used whether you want to receive from a single partition or from all partitions.

In v1:
```python
client = EventHubClient(address)
receiver = client.add_receiver(consumer_group, partition)
client.run()
batch = receiver.receive()
client.stop()
```

In v5:
```python
# Receive
def on_event(partition_context, event):
    print("Received event from partition: {}.".format(partition_context.partition_id))

consumer_client = EventHubConsumerClient.from_connection_string(conn_str, consumer_group, eventhub_name=eh_name)
with consumer_client:

>     consumer_client.receive(on_event=on_event)

Reviewer: From a migration guide perspective, in v1 a receiver is only targeted at one partition,
however in v5 if no partition id is set, then consumer would receive events from all partition.

So I'm thinking for the sake of parity with v1 receiver case, in v5 we also receive events from a single partition.

```suggestion
    consumer_client.receive(on_event=on_event, partition_id=partition_id)
```

# Receive batch
def on_event_batch(partition_context, event_batch):
    print("Partition {}, Received count: {}".format(partition_context.partition_id, len(event_batch)))
    
consumer_client = EventHubConsumerClient.from_connection_string(conn_str, consumer_group, eventhub_name=eh_name)
with consumer_client:

>     consumer_client.receive_batch(on_event_batch=on_event_batch)

Reviewer: same as above
```suggestion
    consumer_client.receive_batch(on_event_batch=on_event_batch, partition_id=partition_id)
```

through documentation or exploring namespace for the types that you can instantiate.

#### Consistency
We now have methods with similar names, signature and location for sending and receiving.
This provides consistency and predictability on the various features of the library.

### Client constructors

- While we continue to support connection strings when constructing a client, the main difference is when using Azure Active Directory.
We now use the new [`azure-identity`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md) library
to share a single authentication solution between clients of different Azure services.

In v1:
```python
# Authenticate with address (full URI string - optionally includes URL-encoded access policy and key). For example:
# "amqps://<URL-encoded-SAS-policy>:<URL-encoded-SAS-key>@<namespace>.servicebus.windows.net/<eventhub-name>"

> eventhub_client = EventHubClient(address)

Reviewer 3: I was talking with @swathipil offline about this one. If I put myself in a customer shoes and I was using address, I would want to see it acknowledge in the migration guide, but we might do a more explicit sentence to say it's not supported anymore.
But ignoring him, I think will just create even more confusion, given we can't know how popular this constructor was.

Thoughts?

> eventhub_client = EventHubClient(address)

Author: I don't think the address was constructed programatically, since the [samples](https://github.com/Azure/azure-sdk-for-python/blob/release/eventhub-v1/sdk/eventhub/azure-eventhubs/examples/test_examples_eventhub.py) don't show this either.
In [this](https://github.com/Azure/azure-sdk-for-python/blob/release/eventhub-v1/sdk/eventhub/azure-eventhubs/examples/batch_send.py) example, it is stored in an env variable, so I seems that it could have been copied from the portal. However, when I check in the portal, I can't find this address anywhere. So, I cannot safely assume that it's copied. 

Based on what Laurent and I discussed, we also don't have telemetry on what the common way used by people was. So, to ensure that customers who used address in v1 are also acknowledged, I could update the address section to this:
```suggestion 
v1:
# Authenticate with address
eventhub_client = EventHubClient(address)
...

v5:
# Address is no longer used for authentication.

#Authenticate with connection string
...
```
By mentioning that address is no longer used in v5, it may not be necessary to include the address format that would be distracting in v1. I can also change this as needed.

> eventhub_client = EventHubClient(address)

Reviewer 2: Lets try that @swathipil

# Authenticate with connection string
eventhub_client = EventHubClient.from_connection_string(conn_str) 

# Authenticate with EventProcessorHost
from azure.eventprocessorhost import (
    AbstractEventProcessor,
    EventHubConfig,
    AzureStorageCheckpointLeaseManager,
    EventProcessorHost)

class EventProcessor(AbstractEventProcessor):
    # Methods for opening connection, processing events, closing connection
    ...

eh_config = EventHubConfig(eh_namespace, eventhub_name, user, key, consumer_group="$default")
storage_manager = AzureStorageCheckpointLeaseManager(storage_account_name, storage_key, lease_container_name)

> host = EventProcessorHost(EventProcessor, eh_config, storage_manager)

Author: It seems there was no way to create EPH with a connection string.
[This ](https://github.com/Azure/azure-sdk-for-python/blob/3d578c002a31f071922d68fd41be9ab4dbe652a2/sdk/eventhub/azure-eventhubs/azure/eventprocessorhost/eph.py#L18 )was the only constructor for EPH in Python.
No other methods to mirror [this](https://github.com/Azure/azure-sdk-for-js/blob/7cd64401a4532716b78c20c0d1005b8e5c81628e/sdk/eventhub/event-processor-host/src/eventProcessorHost.ts#L176) in JS

In v1:
```python
client = EventHubClient(address)
sender = client.add_sender()
client.run()
sender.send(EventData('Single message'))
client.stop()
```

In v5:
```python
producer_client = EventHubProducerClient.from_connection_string(conn_str, eventhub_name)

# Send EventDataBatch
event_data_batch = producer.create_batch()
event_data_batch.add(EventData('Single message'))
producer.send_batch(event_data_batch)

# Send list of EventData
event_data_batch = [EventData('Single message')]
producer.send_batch(event_data_batch)
```

### Receiving events 

- The `run` and `stop` methods were previously used since the single `EventHubClient` controlled the lifecycle for all senders and receivers. In v5, the `run` and `stop` methods are deprecated since the `EventHubConsumerClient` controls its own lifecycle.
- The `add_receiver` method is no longer used to create receiver clients. Instead, the `EventHubConsumerClient` is used for receiving events.

> - The old `receive` method returned a list of `EventData`.
> - The new `receive` calls the user callback `on_event` to process single events for easier and more clear interaction with event data when dealing with multiple partitions. 
> - The new `receive_batch` calls the user callback `on_event_batch` to process batches of events for easier and more clear interaction with event data when dealing with multiple partitions.

Reviewer 2: ```suggestion
```suggestion
- In v1, the `receive` method returned a list of `EventData`. You would call this method repeatedly every time you want receive a set of events. In v5, the new `receive` method takes user callback to process events and any resulting errors. This way, you call the method once and it continues to process incoming events until you stop it.
- Additionally, we have a method `receive_batch` which behaves the same as `receive`, but calls the user callback with a batch of events instead of single events. 
- The same methods can be used whether you want to receive from a single partition or from all partitions.
```

through documentation or exploring namespace for the types that you can instantiate.

Similarly, by having a single entry point for receiving of any type (from single partition, all partitions, or with load balancing and checkpointing features) within Event Hubs, the `EventHubConsumerClient` helps with the discoverability of the API as you can explore all available features for receiving events through methods from a single client, as opposed to searching
through documentation or exploring namespace for the types that you can instantiate.

#### Consistency
We now have methods with similar names, signature and location for sending and receiving.
This provides consistency and predictability on the various features of the library.

### Client constructors

- While we continue to support connection strings when constructing a client, the main difference is when using Azure Active Directory.
We now use the new [`azure-identity`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md) library
to share a single authentication solution between clients of different Azure services.

> In v1:

Reviewer 4: i believe there's some value in providing an inline sample wherever available for each of the v5 versions.

Author: I think a table would have worked. However, since I expanded the constructor samples a little, an inline sample seems like it would be a little messy? Would be up for discussing after I update with the eph sample as well for receiving.

```
In v5:
```python
    # Authenticate with connection string
    producer_client = EventHubProducerClient.from_connection_string(conn_str) 
    consumer_client = EventHubConsumerClient.from_connection_string(conn_str) 

    # Authenticate with Active Directory
    from azure.identity import EnvironmentCredential
    producer_client = EventHubProducerClient(fully_qualified_namespace, eventhub_name, credential=EnvironmentCredential())
    consumer_client = EventHubConsumerClient(fully_qualified_namespace, eventhub_name, consumer_group='$Default', credential=EnvironmentCredential())

    # Authenticate consumer with connection string and checkpoint
    from azure.eventhub.extensions.checkpointstoreblob import BlobCheckpointStore
    checkpoint_store = BlobCheckpointStore.from_connection_string(storage_conn_str, container_name)
    consumer_client = EventHubConsumerClient.from_connection_string(conn_str, consumer_group='$Default', checkpoint_store=checkpoint_store)
```
### Sending events
* `add_sender`, `run`, and `stop` methods are replaced by `from_connection_string` method on `EventGridProducerClient` to more approachably open a connection ready for sending.

> * `send` method is replaced by `send_batch` method on `EventGridProducerClient` to clarify that, instead of single `EventData`, either an `EventDataBatch` or list of `EventData` are sent in one call not exceeding the event hub frame size limit.
> * `EventDataBatch` is created using the `create_batch` method and `EventData` messages are added to the batch using the `add` method, until the size limit is reached.

Reviewer 2: This does not clarify why we want to discourage the sending of single events.

```suggestion
* The `send` method that allowed sending single events in each call is removed in favor of the `send_batch` to encourage sending events in batches for better throughput.
* The new `send_batch` method takes a list of `EventData` objects that is batched into a single message by the client before sending.
* The above approach fails if the list of events increase the size limit of the message. To safely send within size limits, use the `EventDataBatch` object to which you can add `EventData` objects until the size limit is reached after which you can send it using the same `send_batch` method.
``````

In v1:
```python
    client = EventHubClient(address)
    sender = client.add_sender()
    client.run()
    sender.send(EventData('Single message'))
    client.stop()
```

In v5:
```python
    producer_client = EventHubProducerClient.from_connection_string(conn_str, eventhub_name)

    # Send EventDataBatch
    event_data_batch = producer.create_batch()
    event_data_batch.add(EventData('Single message'))
    producer.send_batch(event_data_batch)

    # Send list of EventData
    event_data_batch = [EventData('Single message')]
    producer.send_batch(event_data_batch)
```

### Receiving events 
* `add_receiver`, `run`, and `stop` methods are replaced by `from_connection_string` method on `EventGridConsumerClient` to more approachably open a connection ready for receiving.
* `receive` method is renamed `receive_batch` on `EventGridConsumerClient` to be more consistent in the usage of `batch` suffix in other methods on the producer and consumer when receiving or sending batches.

> * `receive` method on `EventGridConsumerClient` now receives only a single event as opposed to previously receiving a batch of events to more clearly reflect the naming, in which `batch` is not used as a suffix.

Reviewer 2: This suggests that the only major change is in the name of the methods where as the main change is that we now have push mode vs the previous pull mode i.e. now you have to pass callbacks, where as previously you got the events as the output of the method.

In v1:
```python
    client = EventHubClient(address)
    receiver = client.add_receiver(consumer_group, partition)
    client.run()
    batch = receiver.receive()
    client.stop()
```

In v5:
```python
    # Receive
    def on_event(partition_context, event):
        print("Received event from partition: {}.".format(partition_context.partition_id))
    
    consumer_client = EventHubConsumerClient.from_connection_string(conn_str, consumer_group, eventhub_name=eh_name)
    with consumer_client:
        consumer_client.receive(on_event=on_event)

    # Receive batch
    def on_event_batch(partition_context, event_batch):
        print("Partition {}, Received count: {}".format(partition_context.partition_id, len(event_batch)))
        
    consumer_client = EventHubConsumerClient.from_connection_string(conn_str, consumer_group, eventhub_name=eh_name)
    with consumer_client:
        consumer_client.receive_batch(on_event_batch=on_event_batch)
```

### Receiving with checkpoints

> Consuming events and saving checkpoints using a checkpoint store was not available in v1.

Reviewer 2: Well... not true. We had the event processor host that allowed this.

Author: adding the ### Migrating code from `EventProcessorHost` to `EventHubConsumerClient` for receiving events
section back in
